go-controller/pkg/ovn/gateway.go

```diff
@@ -685,3 +688,145 @@ func (gw *GatewayManager) addExternalSwitch(prefix, interfaceID, nodeName, gatew
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete gateway router %s: %w", gw.gwRouterName, err)
+	}
+
+	// Remove external switch
+	err = libovsdbops.DeleteLogicalSwitch(gw.nbClient, gw.extSwitchName)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete external switch %s: %w", gw.extSwitchName, err)
+	}
+
+	exGWexternalSwitch := types.EgressGWSwitchPrefix + gw.extSwitchName
+	err = libovsdbops.DeleteLogicalSwitch(gw.nbClient, exGWexternalSwitch)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete external switch %s: %w", exGWexternalSwitch, err)
+	}
+
+	// This will cleanup the NodeSubnetPolicy in local and shared gateway modes. It will be a no-op for any other mode.
+	gw.delPbrAndNatRules(gw.nodeName, nil)
+	return nil
+}
+
+func (gw *GatewayManager) delPbrAndNatRules(nodeName string, lrpTypes []string) {
+	// delete the dnat_and_snat entry that we added for the management port IP
+	// Note: we don't need to delete any MAC bindings that are dynamically learned from OVN SB DB
+	// because there will be none since this NAT is only for outbound traffic and not for inbound
+	mgmtPortName := util.GetK8sMgmtIntfName(gw.netInfo.GetNetworkScopedName(nodeName))
+	nat := libovsdbops.BuildDNATAndSNAT(nil, nil, mgmtPortName, "", nil)
+	logicalRouter := nbdb.LogicalRouter{
+		Name: gw.clusterRouterName,
+	}
+	err := libovsdbops.DeleteNATs(gw.nbClient, &logicalRouter, nat)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		klog.Errorf("Failed to delete the dnat_and_snat associated with the management port %s: %v", mgmtPortName, err)
+	}
+
+	// delete all logical router policies on ovn_cluster_router
+	gw.removeLRPolicies(nodeName, lrpTypes)
+}
+
+func staticRouteCleanup(nbClient libovsdbclient.Client, nextHops []net.IP) {
+	ips := sets.Set[string]{}
+	for _, nextHop := range nextHops {
+		ips.Insert(nextHop.String())
+	}
+	p := func(item *nbdb.LogicalRouterStaticRoute) bool {
+		return ips.Has(item.Nexthop)
+	}
+	err := libovsdbops.DeleteLogicalRouterStaticRoutesWithPredicate(nbClient, types.OVNClusterRouter, p)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		klog.Errorf("Failed to delete static route for nexthops %+v: %v", ips.UnsortedList(), err)
+	}
+}
+
+// policyRouteCleanup cleansup all policies on cluster router that have a nextHop
+// in the provided list.
+// - if the LRP exists and has the len(nexthops) > 1: it removes
+// the specified gatewayRouterIP from nexthops
+// - if the LRP exists and has the len(nexthops) == 1: it removes
+// the LRP completely
+func (gw *GatewayManager) policyRouteCleanup(nextHops []net.IP) {
+	for _, nextHop := range nextHops {
+		gwIP := nextHop.String()
+		policyPred := func(item *nbdb.LogicalRouterPolicy) bool {
+			for _, nexthop := range item.Nexthops {
+				if nexthop == gwIP {
```

Comment:
You're right. I can modify the predicate to work only for the network managed by this GW manager, since I've set the external IDs in the logical router policy.

Good catch.